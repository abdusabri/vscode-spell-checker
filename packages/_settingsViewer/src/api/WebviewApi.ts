
export interface BaseMessage {
    data: any;
}

export type MessageListener = (e: BaseMessage) => any;

export interface WebviewApi {
    postMessage(msg: any): WebviewApi;
    onmessage: MessageListener | undefined;
}

