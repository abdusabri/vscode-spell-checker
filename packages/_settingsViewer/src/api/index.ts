export * from './message';
export * from './MessageBus';