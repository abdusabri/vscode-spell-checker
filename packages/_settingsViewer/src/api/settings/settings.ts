
export type LocalId = string;
export type LocalList = LocalId[];

export type FileType = string;
export type FileTypeList = FileType[];

export type ConfigTarget = keyof SettingByConfigTarget<void>;

export type Inherited<T> = T | undefined;

export interface SettingByConfigTarget<T> {
    user: T;
    workspace: T;
    folder: T;
    file: T;
}

export interface LocalSetting extends SettingByConfigTarget<LocalList | undefined> {}

export interface Configs extends SettingByConfigTarget<Config | undefined> {};

export interface DictionaryEntry {
    name: string;
    locals: LocalList;
    fileTypes: FileTypeList;
    description?: string;
}

export interface Settings {
    locals: LocalSetting;
    dictionaries: DictionaryEntry[];
    configs: Configs;
}

export interface Config {
    locals: Inherited<LocalList>;
    fileTypesEnabled: Inherited<FileTypeList>;
}

const targetConst = Object.freeze<SettingByConfigTarget<ConfigTarget>>({
    user: 'user',
    workspace: 'workspace',
    folder: 'folder',
    file: 'file',
});

export const configTargets = Object.freeze(Object.keys(targetConst) as ConfigTarget[]);

const setOfConfigTargets = new Set<string>(configTargets);

export function isConfigTarget(target: string | undefined): target is ConfigTarget {
    return target !== undefined && setOfConfigTargets.has(target);
}

// Define the order in which configuration is applied.
export const configTargetToIndex = Object.freeze<SettingByConfigTarget<number>>({
    user: 0,
    workspace: 1,
    folder: 2,
    file: 3,
});


export const configTargetOrder = Object.freeze(Object.entries(configTargetToIndex).sort((a, b) => a[1] - b[1]).map(a => a[0]) as ConfigTarget[]);

/*
locals
filetypes

*/

