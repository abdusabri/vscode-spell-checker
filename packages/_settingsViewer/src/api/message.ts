import { Settings, ConfigTarget } from './settings';
export type Commands = 'ConfigurationChangeMessage'
| 'EnableLanguageIdMessage'
| 'EnableLocalMessage'
| 'RequestConfigurationMessage'
| 'SelectFileMessage'
| 'SelectFolderMessage'
| 'SelectTabMessage'
;

export interface Message {
    command: Commands;
}

export type Messages =
    ConfigurationChangeMessage
    | EnableLanguageIdMessage
    | EnableLocalMessage
    | RequestConfigurationMessage
    | SelectFileMessage
    | SelectFolderMessage
    | SelectTabMessage
    ;

export function isMessage(data: any): data is Message {
    return !!(data
        && typeof data === 'object'
        && data.hasOwnProperty('command')
        && typeof data['command'] === 'string'
    );
}

function isA<T extends Message> (cmd: T['command'], fields: ([keyof T, (v: any) => boolean])[]): (msg: Message) => msg is T {
    return function (msg: Message): msg is T {
        const t = msg as T;
        return msg.command === cmd
        && fields.reduce((success, [key, fn]) => success && fn(t[key]), true);
    };
}

export interface ConfigurationChange {
    activeTab?: string;
    settings: Settings;
}

export interface ConfigurationChangeMessage extends Message {
    command: 'ConfigurationChangeMessage';
    value: ConfigurationChange;
}

export interface RequestConfigurationMessage extends Message {
    command: 'RequestConfigurationMessage';
}

export interface SelectTabMessage extends Message {
    command: 'SelectTabMessage';
    value: string;
}

export interface SelectFolderMessage extends Message {
    command: 'SelectFolderMessage';
    value: string;
}

export interface SelectFileMessage extends Message {
    command: 'SelectFileMessage';
    value: string;
}

export interface EnableLanguageIdMessage extends Message {
    command: 'EnableLanguageIdMessage';
    value: {
        target?: ConfigTarget;
        uri?: string;
        languageId: string;
        enable: boolean;
    };
}

export interface EnableLocalMessage extends Message {
    command: 'EnableLocalMessage';
    value: {
        target: ConfigTarget;
        uri: string | undefined;
        local: string;
        enable: boolean;
    };
}

export const isConfigurationChangeMessage = isA<ConfigurationChangeMessage>('ConfigurationChangeMessage', [['value', isObject]]);
export const isEnableLanguageIdMessage = isA<EnableLanguageIdMessage>('EnableLanguageIdMessage', [['value', isObject]]);
export const isEnableLocalMessage = isA<EnableLocalMessage>('EnableLocalMessage', [['value', isObject]]);
export const isRequestConfigurationMessage = isA<RequestConfigurationMessage>('RequestConfigurationMessage', []);
export const isSelectFileMessage = isA<SelectFileMessage>('SelectFileMessage', [['value', isString]]);
export const isSelectFolderMessage = isA<SelectFolderMessage>('SelectFolderMessage', [['value', isString]]);
export const isSelectTabMessage = isA<SelectTabMessage>('SelectTabMessage', [['value', isString]]);

function isObject(v: any): v is object { return typeof v === 'object' && v !== null; }
function isString(v: any): v is string { return typeof v === 'string'; }
