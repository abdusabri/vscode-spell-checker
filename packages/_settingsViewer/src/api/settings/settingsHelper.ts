import { Config, Configs, configTargets, ConfigTarget } from './settings';

export interface ExtractConfigResult<T> {
    target: ConfigTarget;
    config: T;
}

export function extractConfig<K extends keyof Config>(configs: Configs, key: K): ExtractConfigResult<Config[K]> {

    for (let i = configTargets.length - 1; i >= 0; i--) {
        const target = configTargets[i];
        if (configs[target] && configs[target]![key]) {
            return {
                target,
                config: configs[target]![key],
            };
        }
    }

    return {
        target: 'user',
        config: undefined,
    };
}
