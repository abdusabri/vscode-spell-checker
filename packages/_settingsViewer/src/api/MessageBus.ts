import { WebviewApi } from './WebviewApi';
import { Message, Commands, isMessage, Messages } from './message';

export interface Listener {
    cmd: Commands;
    fn: (message: Message) => any;
    dispose(): void;
}

export class MessageBus {
    protected listeners = new Map<Commands, Set<Listener>>();

    constructor(readonly vsCodeApi: WebviewApi) {
        this.vsCodeApi.onmessage = (msg: MessageEvent) => this.respondToMessage(msg);
    }

    listenFor<M extends Messages>(cmd: M['command'], fn: (message: M) => any): Listener {
        const listener = {
            fn,
            cmd,
            dispose: () => this.listeners.has(cmd) && this.listeners.get(cmd)!.delete(listener),
        }

        this.listeners.set(cmd, this.listeners.get(cmd) || new Set());
        const cmdListeners = this.listeners.get(cmd)!;
        cmdListeners.add(listener);

        return listener;
    }

    postMessage<M extends Messages>(msg: M) {
        this.vsCodeApi.postMessage(msg);
    }

    private respondToMessage(msg: MessageEvent) {
        const message = msg.data;

        if (!isMessage(message)) {
            return;
        }

        const listeners = this.listeners.get(message.command);

        if (!listeners) {
            return;
        }

        for (const listener of listeners) {
            listener.fn(message);
        }
    }
}