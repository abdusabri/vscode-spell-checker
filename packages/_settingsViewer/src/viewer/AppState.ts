import {observable, computed} from 'mobx';
import { Settings, ConfigTarget, LocalId, SettingByConfigTarget, Config, Configs, LocalList, WorkspaceFolder, TextDocument, configTargets, ConfigTargets } from '../api/settings/';
import { normalizeCode, lookupCode } from '../iso639-1';
import { compareBy, compareEach } from '../api/utils/Comparable';
import { uniqueFilter } from '../api/utils';
import { Messenger } from '../api';


type Maybe<T> = T | undefined;
type TabTargets = ConfigTarget | 'file' | 'dictionaries' | 'about';

export interface Tab {
    label: string;
    target: TabTargets;
}

const tabs: Tab[] = [
    { label: 'User', target: 'user' },
    { label: 'Workspace', target: 'workspace' },
    { label: 'Folder', target: 'folder' },
    { label: 'File', target: 'file' },
    { label: 'Dictionaries', target: 'dictionaries' },
    { label: 'About', target: 'about' },
];

export interface LanguageInfo {
    code: string;
    name: string;
    dictionaries: string[];
    enabled: boolean;
}

export interface LanguageConfig {
    languages: LanguageInfo[];
    inherited?: ConfigTarget;
}

export interface LanguageConfigs extends SettingByConfigTarget<LanguageConfig> {}

export interface State {
    activeTabName: string;
    settings: Settings;
    tabs: Tab[];
    activeTab: Tab;
    languageConfig: LanguageConfigs;
}

export interface FoundInConfig<T> {
    value: Exclude<T, undefined>,
    target: ConfigTarget
}

type InheritedFromTarget<T> = {
    value: Exclude<T, undefined>,
    target: ConfigTarget;
}


type InheritMembers<T> = {
    [K in keyof T]: InheritedFromTarget<T[K]>;
}

type InheritedConfigs = SettingByConfigTarget<Config>;

export class AppState implements State {
    @observable activeTabName = '';
    @observable settings: Settings = {
        dictionaries: [
        ],
        configs: {
            user: { inherited: {}, locals: [], languageIdsEnabled: [] },
            workspace: { inherited: { locals: 'user', languageIdsEnabled: 'user' }, locals: [], languageIdsEnabled: [] },
            folder: { inherited: { locals: 'user', languageIdsEnabled: 'user' }, locals: [], languageIdsEnabled: [] },
            file: undefined,
        }
    };
    @observable debugMode: boolean = false;

    constructor(private messageBus: Messenger) {}

    @computed get tabs() {
        const hidden = new Set<TabTargets>(configTargets.filter(target => !this.settings.configs[target]));
        if (this.workspaceFolders.length <= 1) {
            hidden.add(ConfigTargets.folder);
        }
        return tabs.filter(tab => !hidden.has(tab.target));
    }

    @computed get activeTab() {
        return this.tabs.find(t => t.label === this.activeTabName) || this.tabs[0];
    }

    @computed get languageConfig(): LanguageConfigs {
        const calcConfig = (target: ConfigTarget): LanguageConfig => {
            const config = this.settings.configs[target];
            const locals = config.locals; // todo: calc inheritance
            const inherited = config.inherited.locals;

            const infos = new Map<string, LanguageInfo>();

            const addLocalsToInfos = (locals: string[], dictionaryName: string | undefined) => {
                locals.map(normalizeCode).map(lookupCode).filter(notUndefined).forEach(lang => {
                    const { code, lang: language, country } = lang;
                    const name = country ? `${language} - ${country}` : language;
                    const found = this.isLocalEnabledEx(target, code);
                    const enabled = found && found.value || false;
                    const info: LanguageInfo = infos.get(name) || {
                        code,
                        name,
                        enabled,
                        dictionaries: [],
                    };
                    if (dictionaryName) {
                        info.dictionaries.push(dictionaryName);
                    }
                    infos.set(name, info);
                });
            }
            if (locals) {
                addLocalsToInfos(locals, undefined);
            }
            this.settings.dictionaries.forEach(dict => addLocalsToInfos(dict.locals, dict.name));

            return {
                languages: [...infos.values()].sort(compareEach(
                    compareBy(info => !info.dictionaries.length),
                    compareBy('name'),
                )),
                inherited
            };
        };

        return {
            user: calcConfig('user'),
            workspace: calcConfig('workspace'),
            folder: calcConfig('folder'),
        }
    }

    @computed get activeTabIndex(): number {
        const index = this.tabs.findIndex(t => t.label === this.activeTabName);
        return index > 0 ? index : 0;
    }

    @computed get workspaceFolderNames(): string[] {
        const workspace = this.settings.workspace;
        const folders = workspace && workspace.workspaceFolders || [];
        return folders.map(f => f.name);
    }

    @computed get activeWorkspaceFolder(): string | undefined {
        const folder = this.findMatchingFolder(this.settings.activeFolderUri);
        return folder && folder.name;
    }

    @computed get workspaceFolders(): WorkspaceFolder[] {
        const workspace = this.settings.workspace;
        return workspace && workspace.workspaceFolders || [];
    }

    @computed get activeFileUri(): string | undefined {
        return this.settings.activeFileUri;
    }

    @computed get documents(): TextDocument[] {
        const workspace = this.settings.workspace;
        return workspace && workspace.textDocuments || [];
    }

    @computed get documentSelection(): { label: string, value: string }[] {
        return this.documents.map(doc => ({ label: doc.fileName, value: doc.uri }));
    }

    private findMatchingFolder(uri: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter(f => f.uri === uri)[0];
    }

    private findMatchingFolderByName(name: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter(f => f.name === name)[0];
    }

    actionSetLocal(target: ConfigTarget, local: LocalId, enable: boolean) {
        const localsCurrent = this.settings.configs[target].locals;
        const locals = (enable ? [...localsCurrent, local] : localsCurrent.filter(v => v !== local)).filter(uniqueFilter());
        this.settings.configs[target].locals = locals;
        this.messageBus.postMessage({ command: 'EnableLocalMessage', value: { target, local, enable } });
    }

    actionSetDebugMode(isEnabled: boolean) {
        this.debugMode = isEnabled;
    }

    private isLocalEnabledEx(field: ConfigTarget, code: LocalId):InheritedFromTarget<boolean> {
        const config = this.settings.configs[field];
        const target = config.inherited.locals || field;
        const locals = config.locals;
        return  {
            value: locals.map(normalizeCode).includes(code),
            target
        };
    }


    actionActivateTabIndex(index: number) {
        const tab = this.tabs[index];
        if (tab) {
            this.actionActivateTab(tab.label);
        }
    }

    actionActivateTab(tabName: string) {
        this.activeTabName = tabName;
        this.messageBus.postMessage({ command: 'SelectTabMessage', value: this.activeTabName });
    }

    actionSelectFolder(folderName: string) {
        const folder = this.findMatchingFolderByName(folderName);
        const folderUri = folder && folder.uri;
        this.settings.activeFolderUri = folderUri;
        if (folderUri !== undefined) {
            this.messageBus.postMessage({ command: 'SelectFolderMessage', value: folderUri })
        }
    }

    actionSelectDocument(documentUri: string) {
        this.messageBus.postMessage({ command: 'SelectFileMessage', value: documentUri });
    }

    actionEnableLanguageId(languageId: string, enable: boolean, target?: ConfigTarget) {
        const fileConfig = this.settings.configs.file;
        if (fileConfig && fileConfig.languageId === languageId) {
            fileConfig.languageEnabled = enable;
        }
        this.messageBus.postMessage({ command: 'EnableLanguageIdMessage', value: { languageId, enable, target }});
    }
}

function notUndefined<T>(a : T): a is Exclude<T, undefined> {
    return a !== undefined;
}


