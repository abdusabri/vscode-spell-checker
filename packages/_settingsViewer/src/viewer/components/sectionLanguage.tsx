import * as React from 'react';
import {observer} from 'mobx-react';
import { AppState } from '../AppState';
import { ConfigTarget } from '../../api/settings';
import List, { ListItem, ListItemGraphic, ListItemText, ListItemMeta } from '@material/react-list';
import MaterialIcon from '@material/react-material-icon';
import { Checkbox } from '@material/react-checkbox';

function initRipple(){}

@observer
export class SectionLanguage extends React.Component<{appState: AppState, target: ConfigTarget}, {}> {
    render() {
        const handleSelect = (index) => this.handleSelect(index);
        const target = this.props.target;
        const langConfig = this.props.appState.languageConfig[target];
        if (!langConfig) {
            return <div></div>
        }
        const inherited = langConfig.inherited;
        const note = inherited && inherited !== target ? <span style={{ fontSize: '0.65em', opacity: 0.5}}>inherited from {inherited}</span> : '';
        return (
            <div>
                <h3>Language {note}</h3>
                <div>
                    <List twoLine handleSelect={handleSelect}>
                        {langConfig.languages.map(entry => {
                            const hasLocals = entry.dictionaries && entry.dictionaries.length > 0;
                            const icon = hasLocals ? 'import_contacts' : 'block';
                            const subText = entry.dictionaries.join(', ') || 'no dictionaries found';
                            return (
                            <ListItem key={entry.name} role='checkbox'>
                                <ListItemGraphic graphic={<MaterialIcon icon={icon}/>} />
                                <ListItemText primaryText={entry.name} secondaryText={subText} />
                                <ListItemMeta meta={<Checkbox checked={entry.enabled} initRipple={initRipple} />}/>
                            </ListItem>);

                        })}
                    </List>
                </div>
            </div>
        );
     }

     handleSelect(index: number) {
        const appState = this.props.appState;
        const target = this.props.target;
        const langConfig = appState.languageConfig[target];
        console.log(`handelSelect ${index} target ${target}`);
        if (!langConfig) return;
        const langs = langConfig.languages;
        if (!langs) return;
        const lang = langs[index];
        if (!lang) return;
        this.props.appState.setLocal(target, lang.code, !lang.enabled);
    }
}
