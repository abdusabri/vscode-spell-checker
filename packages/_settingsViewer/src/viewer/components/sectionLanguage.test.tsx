
import * as React from 'react';
import { AppState } from '../AppState';
import { create } from 'react-test-renderer';
import { sampleSettings } from '../samples/sampleSettings';
import { SectionLanguage } from './sectionLanguage';
import { configTargets } from '../../api/settings';
import { sampleAppState } from '../../test/fixtures/AppState';

describe('SectionLanguage Verification', () => {
    it('tests the snapshots', () => {
        const appState = getSampleAppState();
        configTargets.forEach(target => {
            const panelRenderer = create(<SectionLanguage appState={appState} target={target}></SectionLanguage>).toJSON()!;
            expect(panelRenderer).toMatchSnapshot(`<SectionLanguage> for target: ${target}`);
        });
    });

    function getSampleAppState(): AppState {
        const appState: AppState = sampleAppState();
        appState.settings = sampleSettings;
        return appState;
    }
});
