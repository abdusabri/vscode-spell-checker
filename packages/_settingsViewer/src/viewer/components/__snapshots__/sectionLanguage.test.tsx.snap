// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionLanguage Verification tests the snapshots: <SectionLanguage> for target: file 1`] = `
<div>
  <h3>
    Language 
    <span
      style={
        Object {
          "fontSize": "0.65em",
          "opacity": 0.5,
        }
      }
    >
      inherited from 
      workspace
    </span>
  </h3>
  <div>
    <ul
      aria-orientation="vertical"
      className="mdc-list mdc-list--two-line"
    >
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={0}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            Czech
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            cs-cz
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en_us
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English - United Kingdom
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en-gb
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English - United States
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en_us
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          block
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            Danish
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            no dictionaries found
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box
        </i>
      </li>
    </ul>
  </div>
</div>
`;

exports[`SectionLanguage Verification tests the snapshots: <SectionLanguage> for target: folder 1`] = `
<div>
  <h3>
    Language 
    <span
      style={
        Object {
          "fontSize": "0.65em",
          "opacity": 0.5,
        }
      }
    >
      inherited from 
      workspace
    </span>
  </h3>
  <div>
    <ul
      aria-orientation="vertical"
      className="mdc-list mdc-list--two-line"
    >
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={0}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            Czech
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            cs-cz
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en_us
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English - United Kingdom
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en-gb
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English - United States
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en_us
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          block
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            Danish
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            no dictionaries found
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box
        </i>
      </li>
    </ul>
  </div>
</div>
`;

exports[`SectionLanguage Verification tests the snapshots: <SectionLanguage> for target: user 1`] = `
<div>
  <h3>
    Language 
    
  </h3>
  <div>
    <ul
      aria-orientation="vertical"
      className="mdc-list mdc-list--two-line"
    >
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={0}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            Czech
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            cs-cz
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en_us
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English - United Kingdom
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en-gb
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English - United States
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en_us
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
    </ul>
  </div>
</div>
`;

exports[`SectionLanguage Verification tests the snapshots: <SectionLanguage> for target: workspace 1`] = `
<div>
  <h3>
    Language 
    
  </h3>
  <div>
    <ul
      aria-orientation="vertical"
      className="mdc-list mdc-list--two-line"
    >
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={0}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            Czech
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            cs-cz
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en_us
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English - United Kingdom
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en-gb
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          import_contacts
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            English - United States
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            en_us
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box_outline_blank
        </i>
      </li>
      <li
        className="mdc-list-item"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="checkbox"
        tabIndex={-1}
      >
        <i
          className="material-icons mdc-list-item__graphic"
          tabIndex={-1}
        >
          block
        </i>
        <span
          className="mdc-list-item__text"
          tabIndex={-1}
        >
          <span
            className="mdc-list-item__primary-text"
            tabIndex={-1}
          >
            Danish
          </span>
          <span
            className="mdc-list-item__secondary-text"
            tabIndex={-1}
          >
            no dictionaries found
          </span>
        </span>
        <i
          className="material-icons mdc-list-item__meta"
          tabIndex={-1}
        >
          check_box
        </i>
      </li>
    </ul>
  </div>
</div>
`;
